hudi-common/src/main/java/org/apache/hudi/common/util/MarkerUtils.java

```diff
@@ -0,0 +1,179 @@
+        return Option.empty();
+      }
+      fsDataInputStream = fileSystem.open(markerTypeFilePath);
+      content = Option.of(MarkerType.valueOf(FileIOUtils.readAsUTFString(fsDataInputStream)));
+    } catch (IOException e) {
+      throw new HoodieIOException("Cannot read marker type file " + markerTypeFilePath.toString()
+          + "; " + e.getMessage(), e);
+    } finally {
+      closeQuietly(fsDataInputStream);
+    }
+    return content;
+  }
+
+  /**
+   * Writes the marker type to the file `MARKERS.type`.
+   *
+   * @param markerType marker type.
+   * @param fileSystem file system to use.
+   * @param markerDir  marker directory.
+   */
+  public static void writeMarkerTypeToFile(MarkerType markerType, FileSystem fileSystem, String markerDir) {
+    Path markerTypeFilePath = new Path(markerDir, MARKER_TYPE_FILENAME);
+    FSDataOutputStream fsDataOutputStream = null;
+    BufferedWriter bufferedWriter = null;
+    try {
+      fsDataOutputStream = fileSystem.create(markerTypeFilePath, false);
+      bufferedWriter = new BufferedWriter(new OutputStreamWriter(fsDataOutputStream, StandardCharsets.UTF_8));
+      bufferedWriter.write(markerType.toString());
+    } catch (IOException e) {
+      throw new HoodieException("Failed to create marker type file " + markerTypeFilePath.toString()
+          + "; " + e.getMessage(), e);
+    } finally {
+      closeQuietly(bufferedWriter);
+      closeQuietly(fsDataOutputStream);
+    }
+  }
+
+  /**
+   * Deletes `MARKERS.type` file.
+   *
+   * @param fileSystem file system to use.
+   * @param markerDir  marker directory.
+   */
+  public static void deleteMarkerTypeFile(FileSystem fileSystem, String markerDir) {
+    Path markerTypeFilePath = new Path(markerDir, MARKER_TYPE_FILENAME);
+    try {
+      fileSystem.delete(markerTypeFilePath, false);
+    } catch (IOException e) {
+      throw new HoodieIOException("Cannot delete marker type file " + markerTypeFilePath.toString()
+          + "; " + e.getMessage(), e);
+    }
+  }
+
+  /**
+   * Reads files containing the markers written by timeline-server-based marker mechanism.
+   *
+   * @param markerDir   marker directory.
+   * @param fileSystem  file system to use.
+   * @param context     instance of {@link HoodieEngineContext} to use
+   * @param parallelism parallelism to use
+   * @return A {@code Map} of file name to the set of markers stored in the file.
+   */
+  public static Map<String, Set<String>> readTimelineServerBasedMarkersFromFileSystem(
+      String markerDir, FileSystem fileSystem, HoodieEngineContext context, int parallelism) {
+    Path dirPath = new Path(markerDir);
+    try {
+      if (fileSystem.exists(dirPath)) {
+        FileStatus[] fileStatuses = fileSystem.listStatus(dirPath);
+        Predicate<String> prefixFilter = pathStr -> pathStr.contains(MARKERS_FILENAME_PREFIX);
+        Predicate<String> markerTypeFilter = pathStr -> !pathStr.equals(MARKER_TYPE_FILENAME);
```

Comment:
I thought you wanted to change this to contains.

Reply:
Right.  I added a more strict check to prevent accidental matching:
```
Predicate<String> markerTypeFilter =
            pathStr -> !stripMarkerFolderPrefix(pathStr, markerDir).equals(MARKER_TYPE_FILENAME);
```